File: services/learn/learn/application/chats/create_chat/create_chat_command.py
import inject
from pydantic import BaseModel, StrictStr


from services.learn.learn.application.chats.dtos.chat_dto import ChatDto
from services.learn.learn.application.chats.chat_mapper import ChatMapper
from services.learn.learn.domain.Chat.entities.chat import Chat
from services.learn.learn.infrastructure.repositories.chat_repository import ChatRepository


class CreateChatCommandDto(BaseModel):
    name: StrictStr
    is_group_chat: bool
    created_by: StrictStr



class CreateChatCommand:
    @inject.autoparams()
    def __init__(self, chat_repository: ChatRepository):
        self.chat_repository = chat_repository

    def execute(self, dto: CreateChatCommandDto) -> ChatDto:
        chat = Chat.create_new_chat(name=dto.name, is_group_chat=dto.is_group_chat, created_by=dto.created_by)
        self.chat_repository.save_or_update_chat(chat)
        return ChatMapper.to_dto(chat)

File: services/learn/Learn.Application/Chats/CreateChat/CreateChatCommand.cs
using Ardalis.Specification.EntityFrameworkCore;
using AutoMapper;
using Learn.Application.Chats.Dtos;
using Learn.Application.Users.Dtos;
using Learn.Application.Users.Specifications;
using Learn.BuildingBlocks.Application.ExecutionContext;
using Learn.Domain.Chats;
using Learn.Domain.Users;
using Learn.Infrastructure.EntityFramework;
using MediatR;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace Learn.Application.Chats.CreateChat
{
    public class CreateChatCommand : IRequest<ChatDto>
    {
        public string Name { get; set; }

        public CreateChatCommand(string name)
        {
            Name = name;
        }
    }

    internal class CreateChatCommandHandler : IRequestHandler<CreateChatCommand, ChatDto>
    {
        private readonly LearnDbContext _dbContext;
        private readonly IChatRepository _chatRepository;
        private readonly IMapper _mapper;
        private readonly IUserContext _userContext;

        public CreateChatCommandHandler(LearnDbContext dbContext, IChatRepository chatRepository, IMapper mapper, IUserContext userContext)
        {
            _dbContext = dbContext;
            _chatRepository = chatRepository;
            _mapper = mapper;
            _userContext = userContext;
        }

        public async Task<ChatDto> Handle(CreateChatCommand request, CancellationToken cancellationToken)
        {
            var createdBy = await _dbContext.Users
                .WithSpecification(new UserAggregateSpecification())
                .WithSpecification(new UserByEmailSpecification(_userContext.Email))
                .FirstAsync();

            var chat = Chat.CreateNew(request.Name, createdBy);

            await _chatRepository.AddAsync(chat);

            return _mapper.Map<ChatDto>(chat);
        }
    }
}
